### PR TITLE
Better error message for all errors

### PR DESCRIPTION
**Pull request by smahmed776**

### CODE REVIEW

File: src/app/(protected)/workspace/[namespace]/settings/general/_components/general-form.tsx [hunk 1]
```diff
@@ -20,28 +20,34 @@ import DeleteWorkspace from "./delete-workspace";
...
-import { Icons } from "~/components/icons";
 import { useWorkspace } from "../../../workspace-provider";
 import { env } from "~/env";
+import { handleTrpcError } from "~/lib/utils";
 import { updateWorkspace } from "~/types/workspace";
+import { workspace } from "~/schemas/public/Workspace";
 
 type Props = {
-  workspaceId: string;
+  workspace: z.infer<typeof workspace>;
```
Comment (itsjoeoui) [inline src/app/(protected)/workspace/[namespace]/settings/general/_components/general-form.tsx:30]: No need to infer here haha, there is a Workspace type in that file

File: src/app/(protected)/workspace/[namespace]/settings/general/page.tsx [hunk 2]
```diff
@@ -16,7 +21,7 @@ async function GeneralPage({ params }: { params: { namespace: string } }) {
         </p>
       </div>
       <Separator />
-      <GeneralForm workspaceId={workspaceId} />
+      {workspace ? <GeneralForm workspace={workspace} /> : <Icons.spinner />}
```
Comment (itsjoeoui) [inline src/app/(protected)/workspace/[namespace]/settings/general/page.tsx:24]: workspace should always have a value here

File: src/app/(verify)/verify/page.tsx [hunk 3]
```diff
@@ -1,10 +1,27 @@
...
+import { toast } from "sonner";
 import { Button } from "~/components/ui/button";
 
 const Page = () => {
   const [countdown, setCountdown] = useState<number>(0);
 
+  const sendEmailVerification = async () => {
+    try {
+      const data = (await axios.post("/api/email-verification", {}))
+        .data as Record<string, string>;
```
Comment (itsjoeoui) [inline src/app/(verify)/verify/page.tsx:13]: most of the time, there should never be something like `as Record<string, string>`, just type it more explicitly.

Edit: I realized you can probably move this into a tRPC route so it is type safe.

File: src/app/api/login/route.ts [hunk 4]
```diff
@@ -37,8 +37,8 @@ export const POST = withAppRouterHighlight(async (request: NextRequest) => {
       .executeTakeFirst();
 
     if (!existingUser) {
-      return new Response("Wrong password or user does not exist!", {
-        status: 400,
+      return new Response("User does not exist!", {
```
Comment (itsjoeoui) [inline src/app/api/login/route.ts:40]: you should kept the old error message
`Wrong password or user does not exist!`, this is a standard practice in the industry. (https://www.gnucitizen.org/blog/username-enumeration-vulnerabilities/)

File: src/app/api/login/route.ts [hunk 5]
```diff
@@ -57,7 +57,7 @@ export const POST = withAppRouterHighlight(async (request: NextRequest) => {
     );
 
     if (!validPassword) {
-      return new Response("Wrong password or user does not exist!", {
+      return new Response("Wrong password!", {
```
Comment (itsjoeoui) [inline src/app/api/login/route.ts:60]: Same as above

Comment (itsjoeoui) [reply src/app/api/login/route.ts:60]: Make sure the error code is also the same, such that people cannot notice the difference

File: src/app/api/password-reset/route.ts [hunk 6]
```diff
@@ -24,27 +24,21 @@ export const POST = withAppRouterHighlight(async (request: NextRequest) => {
       .executeTakeFirst();
 
     if (!storedUser) {
-      return new Response(
-        "A password reset email will be sent if you have a Flojoy Cloud account. Please check your inbox :)",
-        {
-          status: 200,
-        },
-      );
+      return new Response("User does not exist!", {
```
Comment (itsjoeoui) [inline src/app/api/password-reset/route.ts:27]: A password reset email will be sent if you have a Flojoy Cloud account. Please check your inbox :)

Comment (itsjoeoui) [reply src/app/api/password-reset/route.ts:27]: Restore to this old message for the same reason mentioned above

File: src/app/api/signup/route.ts [hunk 7]
```diff
@@ -39,7 +39,18 @@ export const POST = withAppRouterHighlight(async (request: NextRequest) => {
...
     const hashedPassword = await new Argon2id().hash(password);
 
-    // TODO: check if the user already exists
+    const userExists = await db
+      .selectFrom("user")
+      .selectAll()
+      .where("email", "=", parsedEmail.data)
+      .executeTakeFirst();
+
+    if (userExists) {
```
Comment (itsjoeoui) [inline src/app/api/signup/route.ts:48]: Same reason as above, there are 2 ways to handle this.

1. Just log the user in
2. Send user an email saying this account already exists

Make sure the app should behave the same as if it is a normal signup, make a malicious actor not be able to tell apart if this user already has an account or not

File: src/lib/utils.ts [hunk 8]
```diff
@@ -15,3 +16,16 @@ export function bind<T, U>(
 
   return action(val);
 }
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const handleTrpcError = (error: any) => {
+  if (error instanceof TRPCClientError) {
+    if ((error.data as Record<string, string>)?.zodError) {
```
Comment (itsjoeoui) [inline src/lib/utils.ts:23]: `as Record<string, string>` :(

File: src/server/api/routers/hardware.ts [hunk 9]
```diff
@@ -143,92 +145,111 @@ export const hardwareRouter = createTRPCRouter({
...
-            message: "Model not found",
-          });
-        }
-
-        const modelComponents = await getModelComponents(model.id);
-
-        if (modelComponents.length > 0) {
-          const ids = components.map((c) => c.hardwareId);
-          if (_.uniq(ids).length !== ids.length) {
+        try {
```
Comment (itsjoeoui) [inline src/server/api/routers/hardware.ts:150]: should not just wrap the entire thing in a try catch, should handle each error individually :)

File: src/server/api/routers/model.ts [hunk 10]
```diff
@@ -51,38 +52,56 @@ export const modelRouter = createTRPCRouter({
...
-          })
-          .returningAll()
-          .executeTakeFirstOrThrow(
-            () =>
-              new TRPCError({
-                code: "INTERNAL_SERVER_ERROR",
-                message: "Failed to create model",
-              }),
-          );
+        try {
```
Comment (itsjoeoui) [inline src/server/api/routers/model.ts:55]: Same, do not just wrap the entire thing in a try catch

File: src/server/api/routers/project.ts [hunk 11]
```diff
@@ -75,24 +76,41 @@ export const projectRouter = createTRPCRouter({
...
-          })
-          .returningAll()
-          .executeTakeFirstOrThrow(
-            () =>
-              new TRPCError({
-                code: "INTERNAL_SERVER_ERROR",
-                message: "Failed to create project",
-              }),
-          );
+        try {
```
Comment (itsjoeoui) [inline src/server/api/routers/project.ts:79]: Same, do not just wrap the entire thing in a try catch

File: src/server/api/routers/test.ts [hunk 12]
```diff
@@ -52,24 +53,42 @@ export const testRouter = createTRPCRouter({
...
-          })
-          .returningAll()
-          .executeTakeFirstOrThrow(
-            () =>
-              new TRPCError({
-                code: "INTERNAL_SERVER_ERROR",
-                message: "Failed to create test",
-              }),
-          );
+        try {
```
Comment (itsjoeoui) [inline src/server/api/routers/test.ts:56]: Same, do not just wrap the entire thing in a try catch

File: src/server/api/routers/user.ts [hunk 13]
```diff
@@ -40,57 +40,70 @@ export const userRouter = createTRPCRouter({
     .use(workspaceAccessMiddleware)
     .output(z.void())
     .mutation(async ({ ctx, input }) => {
-      const currentWorkspaceUser = await ctx.db
-        .selectFrom("workspace_user")
-        .where("userId", "=", ctx.user.id)
-        .where("workspaceId", "=", ctx.workspace.id)
-        .selectAll()
-        .executeTakeFirst();
+      try {
```
Comment (itsjoeoui) [inline src/server/api/routers/user.ts:43]: Same, do not just wrap the entire thing in a try catch

File: src/server/api/routers/workspace.ts [hunk 14]
```diff
@@ -58,78 +59,99 @@ export const workspaceRouter = createTRPCRouter({
...
-        const newWorkspace = await tx
-          .insertInto("workspace")
-          .values({
-            id: generateDatabaseId("workspace"),
-            ...data,
-            planType: "enterprise",
-          })
-          .returningAll()
-          .executeTakeFirst();
+      try {
```
Comment (itsjoeoui) [inline src/server/api/routers/workspace.ts:62]: Same, do not just wrap the entire thing in a try catch

**Review by itsjoeoui — CHANGES_REQUESTED** [review]

File: src/app/(protected)/workspace/[namespace]/settings/general/_components/general-form.tsx (hunk 1 above)
Comment (smahmed776) [reply src/app/(protected)/workspace/[namespace]/settings/general/_components/general-form.tsx:30]: Didn't notice lol. Thanks

File: src/app/(verify)/verify/page.tsx (hunk 3 above)
Comment (smahmed776) [reply src/app/(verify)/verify/page.tsx:13]: Alright, I've moved this route to tRPC.

File: src/app/api/password-reset/route.ts [hunk 15]
```diff
@@ -16,35 +16,29 @@ export const POST = withAppRouterHighlight(async (request: NextRequest) => {
...
-      status: 400,
-    });
   }
+
+  const token = await createPasswordResetToken(storedUser.id);
+  const resetLink = `${env.NEXT_PUBLIC_URL_ORIGIN}/password-reset/${token}`;
+  await sendPasswordResetLink(storedUser.email, resetLink);
+
+  return new Response(
+    "A password reset email will be sent if you have a Flojoy Cloud account. Please check your inbox :)",
```
Comment (itsjoeoui) [inline src/app/api/password-reset/route.ts:39]: Could u extract to a const string such that people don't accidently update one of them and forgot to update the other.

File: src/app/api/login/route.ts [hunk 16]
```diff
@@ -28,56 +28,50 @@ export const POST = withAppRouterHighlight(async (request: NextRequest) => {
...
+    .executeTakeFirst();
 
-  try {
-    const existingUser = await db
-      .selectFrom("user")
-      .selectAll()
-      .where("email", "=", parsedEmail.data)
-      .executeTakeFirst();
+  if (!existingUser) {
+    return new Response("Wrong password or user does not exist!", {
```
Comment (itsjoeoui) [inline src/app/api/login/route.ts:38]: Could u extract to a const string such that people don't accidently update one of them and forgot to update the other.

File: src/app/api/signup/route.ts [hunk 17]
```diff
@@ -35,60 +32,51 @@ export const POST = withAppRouterHighlight(async (request: NextRequest) => {
     });
   }
 
-  try {
-    const userId = generateDatabaseId("user");
-    const hashedPassword = await new Argon2id().hash(password);
+  const userId = generateDatabaseId("user");
```
Comment (itsjoeoui) [inline src/app/api/signup/route.ts:35]: this should actually be in a try catch, and make sure that it always return the same error, such that you cannot distinguish from "account already exits" and "sign up failed". If the account already exists, you can send an email to that address saying you already have an account.

File: src/env.js [hunk 18]
```diff
@@ -41,7 +41,7 @@ export const env = createEnv({
       .optional()
       // transform to boolean using preferred coercion logic
       .transform((s) => s !== undefined && s !== "false" && s !== "0"),
-    HIGHLIGHT_PROJECT_ID: z.string().default("6gl9mxzg"),
+    HIGHLIGHT_PROJECT_ID: z.string().default("jdkmlr3g"),
```
Comment (itsjoeoui) [inline src/env.js:44]: hum what is this for?

File: src/types/db-utils.ts
Comment (itsjoeoui) [inline src/types/db-utils.ts:1]: should probably not be in `src/types` as this is not a type

**Review by itsjoeoui — CHANGES_REQUESTED** [review]

File: src/env.js (hunk 18 above)
Comment (smahmed776) [reply src/env.js:44]: Ah was created a new project with "Flojoy-cloud" name as previous one has typo in name and could't change the name. I'll revert this anyway. You probably can fix that typo.

File: src/types/db-utils.ts
Comment (smahmed776) [reply src/types/db-utils.ts:1]: Ah right!

File: src/app/api/password-reset/route.ts (hunk 15 above)
Comment (smahmed776) [reply src/app/api/password-reset/route.ts:39]: Sure, thanks!

File: src/app/(public)/login/login-form.tsx [hunk 19]
```diff
@@ -41,16 +42,12 @@ const LoginForm = () => {
...
-        return;
-      }
-      if (error.response?.data && typeof error.response?.data === "string") {
-        toast.error(error.response.data);
-      } else {
-        toast.error(
+      toast.error(
+        handleError(
+          error,
           "Login is currently not available, please try again later :)",
```
Comment (39bytes) [inline src/app/(public)/login/login-form.tsx:48]:
```suggestion
          "Login is currently not available, please try again later :(",
```

File: src/app/(public)/password-reset/[token]/pass-reset-fields.tsx [hunk 20]
```diff
@@ -51,19 +52,13 @@ const PasswordResetFields = ({ token }: PasswordResetFieldsProps) => {
...
-      }
-      if (error.response?.data && typeof error.response?.data === "string") {
-        toast.error(error.response.data);
-      } else {
-        toast.error(
+    onError: (e) =>
+      toast.error(
+        handleError(
+          e,
           "Password reset is currently not available, please try again later :)",
```
Comment (39bytes) [inline src/app/(public)/password-reset/[token]/pass-reset-fields.tsx:59]:
```suggestion
          "Password reset is currently not available, please try again later :(",
```

File: src/app/(public)/password-reset/page.tsx [hunk 21]
```diff
@@ -40,16 +41,12 @@ const PasswordReset = () => {
...
-        return;
-      }
-      if (error.response?.data && typeof error.response?.data === "string") {
-        toast.error(error.response.data);
-      } else {
-        toast.error(
+      toast.error(
+        handleError(
+          error,
           "Password reset is currently not available, please try again later :)",
```
Comment (39bytes) [inline src/app/(public)/password-reset/page.tsx:47]:
```suggestion
          "Password reset is currently not available, please try again later :(",
```

File: src/app/(public)/signup/singup-form.tsx [hunk 22]
```diff
@@ -49,16 +50,12 @@ const SignupForm = () => {
...
-        return;
-      }
-      if (error.response?.data && typeof error.response?.data === "string") {
-        toast.error(error.response.data);
-      } else {
-        toast.error(
+      toast.error(
+        handleError(
+          error,
           "Signup is currently not available, please try again later :)",
```
Comment (39bytes) [inline src/app/(public)/signup/singup-form.tsx:56]:
```suggestion
          "Signup is currently not available, please try again later :(",
```

File: src/app/api/email-verification/[code]/route.ts [hunk 23]
```diff
@@ -39,17 +36,17 @@ export const GET = withAppRouterHighlight(async (req: NextRequest) => {
...
         status: 400,
       });
     }
-    if (!user || user.email !== emailVerification.email) {
-      return new Response(null, {
+    if (user.email !== emailVerification.email) {
+      return new Response(invalidCodeMsg, {
         status: 400,
       });
     }
```
Comment (39bytes) [inline src/app/api/email-verification/[code]/route.ts:52]: why not just condense these all into one condition?

File: src/schemas/public/MeasurementType.ts [hunk 24]
```diff
@@ -6,14 +6,12 @@ import { z } from 'zod';
...
+  | 'dataframe';
 
 export default MeasurementType;
 
 /** Zod schema for measurement_type */
 export const measurementType = z.enum([
   'boolean',
   'dataframe',
-  'scalar',
 ]);;
```
Comment (39bytes) [inline src/schemas/public/MeasurementType.ts:17]: maybe you need to rerun db migrations? migrate all the way down then migrate to latest again

Comment (itsjoeoui) [reply src/schemas/public/MeasurementType.ts]: @smahmed776

**Review by itsjoeoui — APPROVED** [review]

**Pull request merged**